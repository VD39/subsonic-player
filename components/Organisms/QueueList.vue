<script setup lang="ts">
import ArtistsList from '@/components/Atoms/ArtistsList.vue';
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import LinkOrText from '@/components/Atoms/LinkOrText.vue';
import MarqueeScroll from '@/components/Atoms/MarqueeScroll.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import FavouriteButton from '@/components/Molecules/FavouriteButton.vue';
import TrackMeta from '@/components/Molecules/TrackMeta.vue';
import TrackPlayPause from '@/components/Organisms/TrackPlayPause.vue';

defineProps<{
  tracks: QueueTrack[];
}>();

defineEmits(['addToPlaylist', 'playTrack', 'removeFromQueue']);
</script>

<template>
  <template v-if="tracks.length">
    <h4>Tracks: {{ tracks.length }}</h4>

    <div ref="tracksWrapper" class="trackTable">
      <div class="trackHeader">
        <div class="trackCell">Track</div>
        <div class="trackCell trackSecondary">Album/Podcast</div>
        <div class="trackCell trackSecondary">Artists/Author</div>
        <div class="trackCell trackTime">Time</div>
        <div class="trackCell trackOptions" />
      </div>

      <div
        v-for="(track, index) in tracks"
        :key="track.id"
        class="trackRow"
        data-test-id="track"
      >
        <div class="trackCell">
          <div>
            <TrackPlayPause
              :track-id="track.id"
              :image="track.image"
              :track-number="DEFAULT_VALUE"
              @play-track="$emit('playTrack', index)"
            />

            <TrackMeta :track="track" class="trackMeta" />

            <FavouriteButton
              v-if="'favourite' in track"
              :id="track.id"
              :type="track.type"
              :favourite="track.favourite"
            />
          </div>
        </div>

        <div class="trackCell trackSecondary">
          <MarqueeScroll
            v-if="'album' in track && track.album"
            ref="albumMarqueeScroll"
          >
            <LinkOrText
              :is-link="!!track.albumId"
              :text="track.album"
              :to="`/album/${track.albumId}`"
            />
          </MarqueeScroll>

          <LinkOrText
            v-else-if="'podcastName' in track && track.podcastName"
            ref="podcastNameLinkOrText"
            :is-link="!!track.podcastId"
            :text="track.podcastName"
            :to="`/podcast/all/${track.podcastId}`"
          />

          <p v-else ref="albumElse">{{ DEFAULT_VALUE }}</p>
        </div>

        <div class="trackCell trackSecondary">
          <MarqueeScroll
            v-if="'artists' in track && track.artists.length"
            ref="artistsMarqueeScroll"
          >
            <ArtistsList :artists="track.artists" />
          </MarqueeScroll>

          <MarqueeScroll
            v-else-if="'author' in track && track.author"
            ref="authorMarqueeScroll"
          >
            <p>{{ track.author }}</p>
          </MarqueeScroll>

          <p v-else ref="artistsElse">{{ DEFAULT_VALUE }}</p>
        </div>

        <time class="trackCell trackTime">
          {{ track.duration }}
        </time>

        <div class="trackCell trackOptions">
          <ButtonLink
            ref="removeFromQueue"
            icon="PhX"
            title="Remove item from queue"
            icon-weight="bold"
            icon-size="small"
            @click="$emit('removeFromQueue', track.id)"
          >
            Remove item from queue
          </ButtonLink>
        </div>
      </div>
    </div>
  </template>

  <NoMediaMessage
    v-else
    :icon="IMAGE_DEFAULT_BY_TYPE.track"
    message="No tracks found."
  />
</template>
