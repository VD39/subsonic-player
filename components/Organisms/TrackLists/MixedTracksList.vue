<script setup lang="ts">
import ArtistsList from '@/components/Atoms/ArtistsList.vue';
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import LazyLoadContent from '@/components/Atoms/LazyLoadContent.vue';
import LinkOrText from '@/components/Atoms/LinkOrText.vue';
import MarqueeScroll from '@/components/Atoms/MarqueeScroll.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import DropdownDivider from '@/components/Molecules/Dropdown/DropdownDivider.vue';
import DropdownItem from '@/components/Molecules/Dropdown/DropdownItem.vue';
import DropdownMenu from '@/components/Molecules/Dropdown/DropdownMenu.vue';
import FavouriteButton from '@/components/Molecules/FavouriteButton.vue';
import TrackMeta from '@/components/Molecules/TrackMeta.vue';
import TrackPlayPause from '@/components/Organisms/TrackPlayPause.vue';

defineProps<{
  hideRemoveOption?: boolean;
  tracks: MixedTrack[];
}>();

const emit = defineEmits<{
  addToPlaylist: [trackId: string, index: number];
  addToQueue: [track: MixedTrack];
  downloadMedia: [track: MixedTrack];
  dragStart: [track: MixedTrack, event: DragEvent];
  mediaInformation: [track: MixedTrack];
  playTrack: [index: number];
  remove: [
    removeArgs: {
      id: string;
      index: number;
    },
  ];
}>();

const trackHeaderNames = TRACK_HEADER_NAMES.mix;

const hasAddToQueueEvent = computed(
  () => !!getCurrentInstance()?.vnode.props?.onAddToQueue,
);

const hasDragStartEvent = computed(
  () => !!getCurrentInstance()?.vnode.props?.onDragStart,
);

function onDragStart(track: MixedTrack, event: DragEvent) {
  if (!hasDragStartEvent.value) {
    return;
  }

  emit('dragStart', track, event);
}
</script>

<template>
  <div v-if="tracks.length" ref="tracksWrapper" class="trackTable withPreview">
    <div class="trackHeader">
      <div class="trackCell">{{ trackHeaderNames[0] }}</div>
      <div class="trackCell trackSecondary">{{ trackHeaderNames[1] }}</div>
      <div class="trackCell trackSecondary">{{ trackHeaderNames[2] }}</div>
      <div ref="trackTime" class="trackCell trackTime">
        {{ trackHeaderNames[3] }}
      </div>
      <div class="trackCell trackOptions" />
      <div
        v-if="!hideRemoveOption"
        ref="trackRemoveHeader"
        class="trackCell trackOptions"
      />
    </div>

    <LazyLoadContent
      v-for="(track, index) in tracks"
      :key="track.id"
      class="trackRow"
      data-test-id="track"
      :draggable="hasDragStartEvent"
      @dragstart="onDragStart(track, $event)"
    >
      <div class="trackCell">
        <div>
          <TrackPlayPause
            :image="track.image"
            :trackId="track.id"
            :trackNumber="track.trackNumber"
            @playTrack="$emit('playTrack', index)"
          />

          <TrackMeta class="trackMeta" :track />

          <FavouriteButton
            v-if="'favourite' in track"
            :id="track.id"
            :favourite="track.favourite"
            :type="track.type"
          />
        </div>
      </div>

      <div class="trackCell trackSecondary">
        <MarqueeScroll
          v-if="'album' in track && track.album"
          ref="albumMarqueeScroll"
        >
          <LinkOrText
            :isLink="!!track.albumId"
            :text="track.album"
            :to="{
              name: ROUTE_NAMES.album,
              params: {
                [ROUTE_PARAM_KEYS.album.id]: track.albumId,
              },
            }"
          />
        </MarqueeScroll>

        <MarqueeScroll
          v-else-if="'podcastName' in track && track.podcastName"
          ref="podcastNameMarqueeScroll"
        >
          <LinkOrText
            :isLink="!!track.podcastId"
            :text="track.podcastName"
            :to="{
              name: ROUTE_NAMES.podcast,
              params: {
                [ROUTE_PARAM_KEYS.podcast.sortBy]:
                  ROUTE_PODCAST_SORT_BY_PARAMS.All,
                [ROUTE_PARAM_KEYS.podcast.id]: track.podcastId,
              },
            }"
          />
        </MarqueeScroll>

        <p v-else ref="albumElse">{{ DEFAULT_VALUE }}</p>
      </div>

      <div class="trackCell trackSecondary">
        <MarqueeScroll
          v-if="'artists' in track && track.artists.length"
          ref="artistsMarqueeScroll"
        >
          <ArtistsList :artists="track.artists" />
        </MarqueeScroll>

        <MarqueeScroll
          v-else-if="'author' in track && track.author"
          ref="authorMarqueeScroll"
        >
          <p>{{ track.author }}</p>
        </MarqueeScroll>

        <p v-else ref="artistsElse">{{ DEFAULT_VALUE }}</p>
      </div>

      <time class="trackCell trackTime">
        {{ track.formattedDuration }}
      </time>

      <div class="trackCell trackOptions">
        <DropdownMenu>
          <DropdownItem
            v-if="!hideRemoveOption"
            ref="dropdownItemRemove"
            @click="
              $emit('remove', {
                id: track.id,
                index,
              })
            "
          >
            Remove track
          </DropdownItem>
          <DropdownItem
            ref="addToPlaylist"
            @click="$emit('addToPlaylist', track.id, index)"
          >
            Add to playlist
          </DropdownItem>
          <DropdownItem
            ref="mediaInformation"
            @click="$emit('mediaInformation', track)"
          >
            Media information
          </DropdownItem>
          <DropdownItem
            ref="downloadMedia"
            @click="$emit('downloadMedia', track)"
          >
            Download track
          </DropdownItem>
          <DropdownDivider />
          <DropdownItem
            v-if="hasAddToQueueEvent"
            ref="addToQueue"
            @click="$emit('addToQueue', track)"
          >
            Add to queue
          </DropdownItem>
          <DropdownItem ref="playTrack" @click="$emit('playTrack', index)">
            Play Track
          </DropdownItem>
        </DropdownMenu>
      </div>

      <div
        v-if="!hideRemoveOption"
        ref="trackRemoveRow"
        class="trackCell trackOptions"
      >
        <ButtonLink
          ref="removeButton"
          icon="PhX"
          iconSize="small"
          iconWeight="bold"
          title="Remove track"
          @click="
            $emit('remove', {
              id: track.id,
              index,
            })
          "
        >
          Remove track
        </ButtonLink>
      </div>
    </LazyLoadContent>
  </div>

  <NoMediaMessage
    v-else
    :icon="IMAGE_DEFAULT_BY_TYPE.track"
    message="No tracks found."
  />
</template>
