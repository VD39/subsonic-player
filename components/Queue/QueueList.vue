<script setup lang="ts">
import IconButton from '@/components/Buttons/IconButton.vue';
import FavouriteButton from '@/components/Buttons/FavouriteButton.vue';
import TrackDetails from './TrackDetail.vue';

const props = defineProps<{
  currentTrackId: string;
  isCurrentTrackPlaying: boolean;
  trackCanPlay: boolean;
  tracks: QueueTrack[];
}>();

defineEmits(['playCurrentTrack', 'removeItem']);

function isCurrentTrack(id: string) {
  return props.currentTrackId === id;
}
</script>

<template>
  <ul :class="$style.list">
    <li
      v-for="track in tracks"
      :key="track.id"
      :class="[
        $style.listItem,
        {
          [$style.current]: isCurrentTrack(track.id),
        },
      ]"
      data-test-id="track-list-item"
    >
      <TrackDetails
        :track="track"
        :is-current-track="isCurrentTrack(track.id)"
        :is-current-track-playing="isCurrentTrackPlaying"
        :track-can-play="trackCanPlay"
        @play-current-track="$emit('playCurrentTrack', track)"
      />

      <div :class="$style.right">
        <FavouriteButton
          :id="track.id"
          :favourite="track.favourite"
          :type="track.type"
          :icon-size="18"
        />

        <IconButton
          ref="remove"
          icon="PhX"
          title="Remove item from queue"
          icon-weight="bold"
          :icon-size="18"
          @click="$emit('removeItem', track.id)"
        >
          Remove item from queue
        </IconButton>
      </div>
    </li>
  </ul>
</template>

<style module>
.list {
  height: 100%;
  overflow: auto;
}

.listItem {
  @mixin align-center;

  justify-content: space-between;

  &:hover {
    background-color: var(--secondary-background-color);
  }
}

.current {
  background-color: var(--secondary-background-color);
}

.right {
  display: inline-flex;
}
</style>
