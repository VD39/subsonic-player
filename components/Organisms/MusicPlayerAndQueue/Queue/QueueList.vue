<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import MixedTracksList from '@/components/Organisms/TrackLists/MixedTracksList.vue';

const {
  clearQueueList,
  playTrackFromQueueList,
  queueList,
  removeTrackFromQueueList,
} = useAudioPlayer();
const { toggleQueueList } = useQueue();
const { addToPlaylistModal } = usePlaylist();
const { downloadMedia } = useMediaLibrary();
const { openTrackInformationModal } = useMediaInformation();
</script>

<template>
  <div class="queueWrapper">
    <div class="centerAll">
      <ButtonLink
        ref="closeQueueList"
        class="queueAction"
        fullWidth
        :icon="ICONS.queueClose"
        iconWeight="bold"
        title="Close queue list"
        @click="toggleQueueList"
      >
        Close queue list
      </ButtonLink>
    </div>

    <div class="spaceBetween">
      <div />

      <ButtonLink
        ref="clearQueue"
        :icon="ICONS.clear"
        title="Clear queue"
        @click="clearQueueList"
      >
        Clear queue
      </ButtonLink>
    </div>

    <h2>Queue ({{ queueList.length }})</h2>

    <MixedTracksList
      :tracks="queueList"
      @addToPlaylist="addToPlaylistModal"
      @downloadMedia="downloadMedia"
      @mediaInformation="openTrackInformationModal"
      @playTrack="playTrackFromQueueList"
      @remove="({ id }) => removeTrackFromQueueList(id)"
    />
  </div>
</template>
