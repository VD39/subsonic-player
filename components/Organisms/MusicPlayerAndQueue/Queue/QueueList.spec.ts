import type { VueWrapper } from '@vue/test-utils';

import MixedTracksList from '@/components/Organisms/TrackLists/MixedTracksList.vue';
import { useAudioPlayerMock } from '@/test/useAudioPlayerMock';
import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { mount } from '@vue/test-utils';

import QueueList from './QueueList.vue';

const toggleQueueListMock = vi.fn();

mockNuxtImport('useQueue', () => () => ({
  toggleQueueList: toggleQueueListMock,
}));

const {
  clearQueueListMock,
  playTrackFromQueueListMock,
  removeTrackFromQueueListMock,
} = useAudioPlayerMock();

function factory(props = {}) {
  return mount(QueueList, {
    props: {
      ...props,
    },
  });
}

describe('QueueList', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows the MixedTracksList component', () => {
    expect(wrapper.findComponent(MixedTracksList).exists()).toBe(true);
  });

  describe('when MixedTracksList component emits a playTrack event', () => {
    beforeEach(() => {
      wrapper.findComponent(MixedTracksList).vm.$emit('playTrack');
    });

    it('calls the playTrackFromQueueList function', () => {
      expect(playTrackFromQueueListMock).toHaveBeenCalled();
    });
  });

  describe('when MixedTracksList component emits a remove event', () => {
    beforeEach(() => {
      wrapper.findComponent(MixedTracksList).vm.$emit('remove', {
        id: 'id',
      });
    });

    it('calls the removeTrackFromQueueList function', () => {
      expect(removeTrackFromQueueListMock).toHaveBeenCalledWith('id');
    });
  });

  describe('when the close queue list ButtonLink component is clicked', () => {
    beforeEach(() => {
      wrapper.findComponent({ ref: 'closeQueueList' }).vm.$emit('click');
    });

    it('calls the toggleQueueList function', () => {
      expect(toggleQueueListMock).toHaveBeenCalled();
    });
  });

  describe('when the clear queue ButtonLink component is clicked', () => {
    beforeEach(() => {
      wrapper.findComponent({ ref: 'clearQueue' }).vm.$emit('click');
    });

    it('calls the clearQueueList function', () => {
      expect(clearQueueListMock).toHaveBeenCalled();
    });
  });
});
