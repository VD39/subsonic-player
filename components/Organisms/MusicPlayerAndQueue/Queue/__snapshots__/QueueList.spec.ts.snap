// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QueueList > matches the snapshot 1`] = `
"<div class="queueWrapper">
  <div class="centerAll"><button class="centerItems buttonLink centerAll fullWidth queueAction queueAction" draggable="false" type="button" title="Close queue list">
      <phcaretdown aria-hidden="true" class="icon" size="24" weight="bold"></phcaretdown><span class="clamp text visuallyHidden"> Close queue list </span>
    </button></div>
  <div class="spaceBetween">
    <div></div><button class="centerItems buttonLink" draggable="false" type="button" title="Clear queue">
      <phtrash aria-hidden="true" class="icon" size="24" weight="regular"></phtrash><span class="clamp text visuallyHidden"> Clear queue </span>
    </button>
  </div>
  <h2>Queue (1)</h2>
  <div class="trackTable withPreview">
    <div class="trackHeader">
      <div class="trackCell">Track</div>
      <div class="trackCell trackSecondary">Album/Podcast</div>
      <div class="trackCell trackSecondary">Artists/Author</div>
      <div class="trackCell trackTime">Duration</div>
      <div class="trackCell trackOptions"></div>
      <div class="trackCell trackOptions"></div>
    </div>
    <div class="trackRow" data-test-id="track" draggable="false">
      <div class="trackCell">
        <div>
          <div class="overlapWrapper trackPlayPause withImage">
            <div class="centerAll preloadImage overlap preloadImage"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="Image" class="image" draggable="false" src="image" style="display: none;"></div>
            <p class="overlap trackNumber visuallyHidden">1</p>
            <div class="overlap playPauseWrapper"><button class="centerItems buttonLink centerAll themeHoverButton globalLink buttonLink centerAll themeHoverButton globalLink buttonLink" draggable="false" type="button">
                <phplay aria-hidden="true" class="icon" size="24" weight="regular"></phplay><span class="clamp text visuallyHidden"> Play track </span>
              </button></div>
          </div>
          <div class="noTouchEvents trackMeta">
            <div class="marqueeScroll mBXS" inert="">
              <div class="bulletList inner">
                <div class="content">
                  <h4>queue-track-0-name</h4>
                </div>
              </div>
            </div>
            <div class="marqueeScroll smallFont mobileOnly mBXS">
              <div class="bulletList inner">
                <div class="content">
                  <ul class="bulletList">
                    <li><span class="visuallyHidden">Track album</span><a class="link" data-test-id="link">album</a></li>
                    <!--v-if-->
                    <li><span class="visuallyHidden">Track artists</span>
                      <ul class="commaList">
                        <li data-test-id="artist-list-item"><a aria-label="Go to artist name" class="link" draggable="false">name</a></li>
                        <li data-test-id="artist-list-item"><a aria-label="Go to artist name1" class="link" draggable="false">name1</a></li>
                      </ul>
                    </li>
                    <!--v-if-->
                    <li><span class="visuallyHidden">Duration</span><span><!--v-if--><time>02:00</time></span></li>
                  </ul>
                </div>
              </div>
            </div>
          </div><button class="centerItems buttonLink" draggable="false" type="button" title="Like track">
            <phheart aria-hidden="true" class="icon" size="24" weight="regular"></phheart><span class="clamp text visuallyHidden">Like track</span>
          </button>
        </div>
      </div>
      <div class="trackCell trackSecondary">
        <div class="marqueeScroll">
          <div class="bulletList inner">
            <div class="content"><a class="link" data-test-id="link">album</a></div>
          </div>
        </div>
      </div>
      <div class="trackCell trackSecondary">
        <div class="marqueeScroll">
          <div class="bulletList inner">
            <div class="content">
              <ul class="commaList">
                <li data-test-id="artist-list-item"><a aria-label="Go to artist name" class="link" draggable="false">name</a></li>
                <li data-test-id="artist-list-item"><a aria-label="Go to artist name1" class="link" draggable="false">name1</a></li>
              </ul>
            </div>
          </div>
        </div>
      </div><time class="trackCell trackTime">02:00</time>
      <div class="trackCell trackOptions">
        <div class="centerItems dropdownMenu"><button class="centerItems buttonLink alignRight buttonLink buttonLink" draggable="false" type="button" title="More">
            <phdotsthreevertical aria-hidden="true" class="icon" size="24" weight="regular"></phdotsthreevertical><span class="clamp text visuallyHidden">More</span>
          </button>
          <!--v-if-->
        </div>
      </div>
      <div class="trackCell trackOptions"><button class="centerItems buttonLink" draggable="false" type="button" title="Remove track">
          <phx aria-hidden="true" class="icon" size="18" weight="bold"></phx><span class="clamp text visuallyHidden"> Remove track </span>
        </button></div>
    </div>
  </div>
</div>"
`;
